### PR TITLE
[SPARK-38623][SQL] Add more comments and tests for HashShuffleSpec

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Add more comments and tests to explain the special handling of duplicated cluster keys.

### Why are the changes needed?

improve code readability

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @sunchao

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 1]
```diff
@@ -94,17 +94,25 @@ case class ClusteredDistribution(
     HashPartitioning(clustering, numPartitions)
   }
 
+  lazy val distKeyToPos = {
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:97]: should we declare the type for public val

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:97]:
```suggestion
  lazy val distKeyToPos: Map[Expression, Int] = {
```

**Comment by sunchao** [comment]
Hmm what if you have cluster keys: `AB` and `AD`, and distribution keys: `ABB` and `ACD`, or cluster keys: `BB` and `CD` and distribution keys: `BB` and `CD`? will the new approach make them invalid?

**Comment by cloud-fan** [comment]
@sunchao you are right, let me put your examples in the tests.

**Review by c21 — APPROVED** [review]
LGTM

**Review by LuciferYang — APPROVED** [review]
LGTM +1

**Review by sunchao — APPROVED** [review]
LGTM, thanks @cloud-fan !

**Comment by cloud-fan** [comment]
thanks for review, merging to master/3.3!

**Pull request closed without merging**